scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/BucketMetadataUtil.java

```diff
@@ -0,0 +1,167 @@
+  }
+
+  public <K, V> SourceMetadata<K, V> getSourceMetadata(
+      List<ResourceId> directories, String filenameSuffix) {
+    final List<FileAssignment> fileAssignments =
+        directories.stream()
+            .sorted(Comparator.comparing(ResourceId::toString).reversed())
+            .map(dir -> getFileAssignment(dir, filenameSuffix))
+            .collect(Collectors.toList());
+
+    final int total = fileAssignments.size();
+    final Map<ResourceId, PartitionMetadata> partitionMetadata = new HashMap<>();
+    BucketMetadata<K, V> canonicalMetadata = null;
+    ResourceId canonicalMetadataDir = null;
+    int start = 0;
+    while (start < total) {
+      final List<FileAssignment> input =
+          fileAssignments.subList(start, Math.min(total, start + batchSize));
+      final List<Optional<BucketMetadata<K, V>>> result =
+          input
+              .parallelStream()
+              .map(BucketMetadataUtil::<K, V>getMetadata)
+              .collect(Collectors.toList());
+
+      if (result.stream().anyMatch(o -> !o.isPresent())) {
+        // Fail fast if any partition is missing metadata
+        return SourceMetadata.create(null, Collections.emptyMap());
+      }
+
+      for (int i = 0; i < result.size(); i++) {
+        final BucketMetadata<K, V> metadata = result.get(i).get();
+        final FileAssignment fileAssignment = input.get(i);
+        final ResourceId dir = fileAssignment.getDirectory();
+
+        if (canonicalMetadata == null) {
+          canonicalMetadata = metadata;
+          canonicalMetadataDir = dir;
+        }
+
+        Preconditions.checkState(
+            metadata.isCompatibleWith(canonicalMetadata)
+                && metadata.isPartitionCompatible(canonicalMetadata),
+            "Incompatible partitions. Metadata %s is incompatible with metadata %s. %s != %s",
+            dir,
+            canonicalMetadataDir,
+            metadata,
+            canonicalMetadata);
+
+        if (metadata.getNumBuckets() < canonicalMetadata.getNumBuckets()) {
+          canonicalMetadata = metadata;
+          canonicalMetadataDir = dir;
+        }
+
+        final PartitionMetadata value =
+            PartitionMetadata.create(
+                fileAssignment, metadata.getNumBuckets(), metadata.getNumShards());
+        partitionMetadata.put(dir, value);
+      }
+
+      start += batchSize;
+    }
+    return SourceMetadata.create(canonicalMetadata, partitionMetadata);
+  }
+
+  ////////////////////////////////////////////////////////////////////////////////
+
+  private static FileAssignment getFileAssignment(ResourceId directory, String filenameSuffix) {
+    return new SMBFilenamePolicy(directory, filenameSuffix).forDestination();
+  }
+
+  private static <K, V> Optional<BucketMetadata<K, V>> getMetadata(FileAssignment fileAssignment) {
+    final ResourceId resourceId = fileAssignment.forMetadata();
+    try {
+      InputStream inputStream = Channels.newInputStream(FileSystems.open(resourceId));
```

Comment:
That `InputStream` is never closed. Is it expected ?

Reply:
IIRC the underlying Jackson `ObjectMapper` closes the stream but let me just put a break point on it and verify. BRB once I get intellij to build.

Reply:
Yep verified that `ObjectMapper` will close it after reading value.

Reply:
Will do in a separate commit since I need to fix up `BucketMetadata.from` a bit.